Add feedback on Parse JSON processor (#9917)
opensearch-project/documentation-website: _data-prepper/pipelines/configuration/processors/parse-json.md

| Option | Required | Type | Description |
| :--- | :--- | :--- | :--- | 
| `source` | No | String | The field in the event that will be parsed. Default is `message`. |

> | `destination` | No | String | The destination field for the parsed JSON. Defaults to the root of the event. Cannot be `""`, `/`, or any white-space-only string. |

Reviewer: > The destination field for the parsed JSON.

Maybe it would be clearer to say:

> The destination field for fields parsed from the JSON.

> | `destination` | No | String | The destination field for the parsed JSON. Defaults to the root of the event. Cannot be `""`, `/`, or any white-space-only string. |

Author: ```suggestion
| `destination` | No | String | The destination field for the parsed JSON.  Default is the root of the event. Cannot be `""`, `/`, or any white-space-only string. |
```

| `parse_when` | No | String | A condition expression that determines when to parse the field. Accepts a string following the [expression syntax]({{site.url}}{{site.baseurl}}/data-prepper/pipelines/expression-syntax/). |
| `overwrite_if_destination_exists` | No | Boolean | Whether to overwrite the destination field if it already exists. Default is `true`. |
| `delete_source` | No | Boolean | Whether to delete the source field after parsing. Default is `false`. |
| `tags_on_failure` | No | String | A list of tags to apply if parsing fails or an unexpected exception occurs. |

## Usage

To use the `parse_json` processor, add it to your `pipeline.yaml` configuration file:

```yaml
parse-json-pipeline:
  source:
    ...
  ...
  processor:
    - parse_json:
```

### Basic example

This example parses a JSON message field and flattens the data into the event.

For example, the following JSON message contains a key-value pair:

```json
{"outer_key": {"inner_key": "inner_value"}}
```
{% include copy.html %}

> In example event, the original `message` field remains, and the parsed content is added at the root level. Use the `delete_source` option if you want to remove the original field:

Reviewer: > In example event, ...

Should this be?

> For the example event above, ...

> In example event, the original `message` field remains, and the parsed content is added at the root level. Use the `delete_source` option if you want to remove the original field:

Reviewer: >  Use the `delete_source` option if you want to remove the original field:

The example below includes the original. So putting this sentence about `delete_source` is a little confusing in the context.

I recommend either of these changes:
* Drop the sentence about `delete_source`; or
* Create a new paragraph below that has this sentence and then show the result (which would not have `message`).

> In example event, the original `message` field remains, and the parsed content is added at the root level. Use the `delete_source` option if you want to remove the original field:

Author: ```suggestion
From the example event, the original `message` field remains, and the parsed content is added at the root level. Use the `delete_source` option if you want to remove the original field:
```

| `overwrite_if_destination_exists` | No | Boolean | Whether to overwrite the destination field if it already exists. Default is `true`. |
| `delete_source` | No | Boolean | Whether to delete the source field after parsing. Default is `false`. |
| `tags_on_failure` | No | String | A list of tags to apply if parsing fails or an unexpected exception occurs. |

## Usage

To use the `parse_json` processor, add it to your `pipeline.yaml` configuration file:

```yaml
parse-json-pipeline:
  source:
    ...
  ...
  processor:
    - parse_json:
```
{% include copy.html %}

All examples use the following JSON message for the event output:

```json
{"outer_key": {"inner_key": "inner_value"}}
```
{% include copy.html %}

### Basic example

This example parses a JSON message field and flattens the data into the event. From the example event, the original `message` field remains, and the parsed content is added at the root level, as shown in the following output: 

```json
{
  "message": "{\"outer_key\": {\"inner_key\": \"inner_value\"}}",
  "outer_key": {
    "inner_key": "inner_value"
  }
}
```

### Delete a source

> If you want to remove the original field from the orginating JSON message, use the `delete_source` option, as shown in the following example pipeline:

Author: ```suggestion
If you want to remove the original field from the originating JSON message, use the `delete_source` option, as shown in the following example pipeline:
```

```yaml
parse-json-pipeline:
  source:
    ...
  ...
  processor:
    - parse_json:
        delete_source: true
```
{% include copy.html %}

> In the following event, the `message` field is parsed and removed from the event, leaving only the structured output:

Reviewer 2: Remove "from the event" (the event is already referenced in the first phrase).

```json
{
  "outer_key": {
    "inner_key": "inner_value"
  }
}
```

### Example using a JSON pointer

You can use the `pointer` option to extract a specific nested field from the JSON data, as shown in the following example pipeline:

```yaml
parse-json-pipeline:
  source:
    ...
  ...
  processor:
    - parse_json:
        pointer: "/outer_key/inner_key"
```
{% include copy.html %}

> Only the value at the pointer path `/outer_key/inner_key` is extracted and added to the event. If you set `destination`, the extracted value will be added under that field instead:

Reviewer 2: Should "under" be "to"?
